lib/model/rwfolder.go

```diff
@@ -369,22 +372,47 @@ func (f *sendReceiveFolder) pullerIteration(ignores *ignore.Matcher) int {
 	// (directories, symlinks and deletes) goes into the "process directly"
 	// pile.
 
-	folderFiles.WithNeed(protocol.LocalDeviceID, func(intf db.FileIntf) bool {
-		if shouldIgnore(intf, ignores, f.IgnoreDelete) {
+	// Don't iterate over invalid/ignored files unless ignores have changed
+	iterate := folderFiles.WithNeedExcludingInvalid
+	if ignoresChanged {
+		iterate = folderFiles.WithNeed
+	}
+
+	iterate(protocol.LocalDeviceID, func(intf db.FileIntf) bool {
+		l.Debugln(f, "pullerIteration iterate over", intf.FileName())
+
+		if f.IgnoreDelete && intf.IsDeleted() {
 			return true
 		}
 
-		if err := fileValid(intf); err != nil {
-			// The file isn't valid so we can't process it. Pretend that we
-			// tried and set the error for the file.
-			f.newError("need", intf.FileName(), err)
+		// If filename isn't valid, we can terminate early with an appropriate error.
+		// in case it is deleted, we don't care about the filename, so don't complain.
+		if !intf.IsDeleted() && runtime.GOOS == "windows" && fs.WindowsInvalidFilename(intf.FileName()) {
+			f.newError("need", intf.FileName(), fs.ErrInvalidFilename)
 			changed++
 			return true
 		}
 
 		file := intf.(protocol.FileInfo)
 
 		switch {
+		case ignores.ShouldIgnore(file.Name):
+			invalidFile := protocol.FileInfo{
```

Comment:
This desperately needs a helper method on FileInfo or so. To be used also in the other one or two places we duplicate and slightly change a fileinfo. Otherwise we are sure to forget some of these when we add fields.

Reply:
Seems likely indeed. One question: The existing code that "invalidates" a FileInfo creates a new object instead of modifying the existing one:
https://github.com/imsodin/syncthing/blob/8cf6c6e94d273471fb715e25abeff152d3221b06/lib/model/model.go#L1975
The original FileInfo variable isn't used later on. Is there a side effect of changing it in place that I am missing or can such a new helper function modify the FileInfo in place like this:

```go
// Invalidate sets the Invalid property to true, discards block information and
// sets ModifiedBy to the ID of the device that invalidates it.
func (f FileInfo) Invalidate(invalidatedBy ShortID) {
	f.Invalid = true
	f.ModifiedBy = invaliatedBy
	f.Blocks = nil
}
```

Reply:
If you don't reuse it later you can modify it in place. I think in other places we do need to modify a *copy*, and a copy isn't trivial to get as there are slices and stuff internally. I think the helper function we really need is a copy function for FileInfos.

Reply:
I didn't find a place where a copy is needed, however it still wasn't that simple as `db.FileInfoTruncated` is involved, so I had to create a separation function for that type. Hardly the most elegant version, but that's what I could come up with.